### PR TITLE
install: Overhaul Dockerfile.amd for MPI support

### PR DESCRIPTION
**Pull request by FabioLuporini**

### CODE REVIEW

File: docker/Dockerfile.amd [hunk 1]
```diff
@@ -2,72 +2,129 @@
...
-ENV OMPI_CXX=$AOMP/bin/amdclang++
-ENV OMPI_F90=$AOMP/bin/flan
-
-# Set mpiexec flags for gpu aware mpi
-RUN echo "--mca btl '^openib' -x UCX_TLS=sm,self,rocm_copy,rocm_ipc --mca pml ucx" > /opt/openmpi/etc/mpiexec.conf
+    apt-get install -y dh-autoreconf python3-venv python3-dev python3-pip
 
 # Build mpi4py against amdclang
-RUN python3 -m venv /venv && \
+RUN which mpicc && \
```
Comment (georgebisbas) [inline docker/Dockerfile.amd:121]: Would it hurt to keep?

**Review by georgebisbas — APPROVED** [review]

File: docker/Dockerfile.amd (hunk 1 above)
Comment (FabioLuporini) [reply docker/Dockerfile.amd:121]: which part?

File: devito/arch/compiler.py [hunk 2]
```diff
@@ -487,24 +487,24 @@ def __init__(self, *args, **kwargs):
...
+                self.ldflags += ['-fopenmp']
+                self.ldflags += ['--offload-arch=%s' % platform.march]
         elif platform in [POWER8, POWER9]:
             # It doesn't make much sense to use AOMP on Power, but it should work
             self.cflags.append('-mcpu=native')
         else:
             self.cflags.append('-march=native')
 
-        # Generic amd flags
-        self.ldflags.extend(['-fopenmp', '-target', 'x86_64-pc-linux-gnu'])
```
Comment (mloubout) [inline devito/arch/compiler.py:501]: I know these can be removed but you still need the other flag `offload-arch`

File: docker/Dockerfile.amd [hunk 3]
```diff
@@ -2,72 +2,129 @@
 # This Dockerfile contains AMD compilers
 ##############################################################
 
+ARG IMAGE="rocm/dev-ubuntu-22.04:5.4.2-complete"
 ARG arch="aomp"
 
-ARG ROCM_VERSION=5.4.2
```
Comment (mloubout) [inline docker/Dockerfile.amd:7]: This seems easier to maintain it's the exact same but can easily switch version and much easier to use

File: devito/arch/compiler.py (hunk 2 above)
Comment (mloubout) [reply devito/arch/compiler.py:501]: Also need to fix the arch detection to use `offload-arch` not `mygpu`

File: docker/Dockerfile.amd [hunk 4]
```diff
@@ -2,72 +2,129 @@
...
-                 --prefix=/opt/openmpi/ --with-ucx=/opt/ucx \
-                 --enable-mca-no-build=btl-uct --enable-mpi1-compatibility && \
-    make -j $(( $(lscpu | awk '/^Socket\(s\)/{ print $2 }') * $(lscpu | awk '/^Core\(s\) per socket/{ print $4 }') )) && \
-    make install && \
-    rm -rf /deps/openmpi
-
-# Set OpenMPI path
-ENV PATH=${PATH}:/opt/openmpi/bin:$AOMP/bin
-ENV LD_LIBRARY_PATH=${LD_LIBRARY_PATH}:/opt/openmpi/lib:$AOMP/lib
-ENV OMPI_CC=$AOMP/bin/amdclang
```
Comment (mloubout) [inline docker/Dockerfile.amd:58]: Pretty sure need those (and hip one for hip)

Comment (mloubout) [inline docker/Dockerfile.amd:58]: Need those (and hip one for hip build)

File: docker/Dockerfile.amd [hunk 5]
```diff
@@ -2,72 +2,129 @@
...
+    && make install
+
+# Install OpenMPI
+RUN cd /tmp \
+    && git clone --recursive https://github.com/open-mpi/ompi.git -b ${OMPI_BRANCH} \
+    && cd ompi \
+    && ./autogen.pl \
+    && mkdir build \
+    && cd build \
+    && ../configure --prefix=$OMPI_HOME --with-ucx=$UCX_HOME \
```
Comment (mloubout) [inline docker/Dockerfile.amd:86]: If you don't build openmpi against amdclang you need to change the amdclang compiler to have `self.mpicc` as amdclang or it's gonna compile with gcc

**Review by mloubout — CHANGES_REQUESTED** [review]
I think there is a few things that need change this looks like a start but will create lots of issues

File: devito/arch/compiler.py (hunk 2 above)
Comment (FabioLuporini) [reply devito/arch/compiler.py:501]: ah OK, probably still using the old image?

Comment (mloubout) [reply devito/arch/compiler.py:501]: Unless you pushed a new one by hand yes it's still using the old one

File: docker/Dockerfile.amd [hunk 6]
```diff
@@ -2,94 +2,170 @@
 # This Dockerfile contains AMD compilers
 ##############################################################
```
Comment (FabioLuporini) [inline docker/Dockerfile.amd:3]: Perhaps we add a reference to https://github.com/amd/InfinityHub-CI/tree/main/base-gpu-mpi-rocm-docker ?

File: docker/Dockerfile.amd (hunk 1 above)
Comment (FabioLuporini) [reply docker/Dockerfile.amd:130]: we can drop `which mpicc` yeah

File: docker/Dockerfile.amd [hunk 7]
```diff
@@ -2,94 +2,170 @@
...
 
+
 ########################################################################
 # HIPCC for GPUs (HIP)
 ########################################################################
 # This will only trigger if arch is hip since the final stage depends on it
 FROM sdk-base as hip
 
+# MPI env
+ENV OMPI_CC="hipcc"
```
Comment (FabioLuporini) [inline docker/Dockerfile.amd:160]: Note: in my experiments, I haven't tried OMPI_CC=hipcc. 

And even if I did, nothing would have changed.

Remember that the HipCompiler class sets mpicc=hipcc already , so we can actually drop these lines and, if we want to, leave a comment about the above

File: docker/Dockerfile.amd [hunk 8]
```diff
@@ -2,94 +2,170 @@
...
+    build-essential \
+    autoconf \
+    automake \
+    libtool \
+    pkg-config \
+    libnuma* \
+    gfortran \
+    flex \
+    hwloc \
+  && apt-get clean
```
[the remark this answers is not in the document]
Comment (mloubout) [reply docker/Dockerfile.amd:38]: Tmux is there in the list and o added back tmpi below already

File: docker/Dockerfile.amd [hunk 9]
```diff
@@ -2,94 +2,170 @@
...
+
+# Until rocm base has it fixed
+RUN ln -s /opt/rocm/llvm/bin/offload-arch /opt/rocm/bin/offload-arch
```
[the remark this answers is not in the document]
Comment (mloubout) [reply docker/Dockerfile.amd:41]: Ok lemme check it they pushed and will remove

File: docker/Dockerfile.amd (hunk 7 above)
Comment (mloubout) [reply docker/Dockerfile.amd:160]: I would keep it for homogeneity in case for example someone uses a fancy version of customcompiler

Comment (FabioLuporini) [reply docker/Dockerfile.amd:160]: OK

File: docker/Dockerfile.amd (hunk 9 above)
[the remark this answers is not in the document]
Comment (mloubout) [reply docker/Dockerfile.amd:41]: Nope not pushed

[the remark this answers is not in the document]
Comment (mloubout) [reply docker/Dockerfile.amd:41]: No that's not fixed it's a hack in their dockerfile the same way we have it here, but it has to be fixed directly into `rocm` and no idea when they'll do that

[the remark this answers is not in the document]
Comment (FabioLuporini) [reply docker/Dockerfile.amd:42]: sorry you're right, when I wrote the stuff above my head was still in INCLUDE+ mode. And clearly "fixed" was poor wording. OK

File: docker/Dockerfile.nvidia [hunk 10]
```diff
@@ -185,7 +187,7 @@ RUN cd /llvm-project/build && \
     -DCLANG_OPENMP_NVPTX_DEFAULT_ARCH=sm_86 \
     -DLIBOMPTARGET_NVPTX_COMPUTE_CAPABILITIES=all \
     -DCMAKE_INSTALL_PREFIX=/llvm && \
-    make -j $(( $(lscpu | awk '/^Socket\(s\)/{ print $2 }') * $(lscpu | awk '/^Core\(s\) per socket/{ print $4 }') )) && \
+    make -j ${nproc} && \
```
Comment (FabioLuporini) [inline docker/Dockerfile.nvidia:190]: ah! what's this?

Comment (mloubout) [reply docker/Dockerfile.nvidia:190]: I think it's a predefined env var by docker makes thing much simpler

**Review by mloubout — APPROVED** [review]

**Pull request merged**